File: java/Candy.java
public class Candy {
    public int candy(int[] ratings) {
        if (ratings == null) {
            return 0;
        }
        int n = ratings.length;
        if (n == 0) {
            return 0;
        }
        int[] lessLeft = new int[n];
        lessLeft[0] = 0;
        for (int i = 1; i < n; ++i) {
            if (ratings[i - 1] < ratings[i]) {
                lessLeft[i] = lessLeft[i - 1] + 1;
            } else {
                lessLeft[i] = 0;
            }
        }
        int[] greaterRight = new int[n];
        greaterRight[n - 1] = 0;
        for (int i = n - 2; i >= 0; --i) {
            if (ratings[i + 1] < ratings[i]) {
                greaterRight[i] = greaterRight[i + 1] + 1;
            } else {
                greaterRight[i] = 0;
            }
        }
        int sum = 0;
        for (int i = 0; i < n; ++i) {
            sum += 1 + Math.max(lessLeft[i], greaterRight[i]);
        }
        return sum;
    }
    
    public static void main(String[] args) {
        Candy solver = new Candy();
        int[] ratings = null;
        ratings = new int[] {1, 2, 2, 2, 3};
        System.out.println(solver.candy(ratings));

        ratings = new int[] {1, 2, 2};
        System.out.println(solver.candy(ratings));
    }
}

File: python/candy.py
#!/usr/bin/env python3

class Solution:
    # @param ratings, a list of integer
    # @return an integer
    def candy(self, ratings):
        res = 0
        n = len(ratings)
        lm = [0] * n
        rm = [0] * n
        for i in range(1, n):
            lm[i] = lm[i - 1] + 1 if ratings[i] > ratings[i - 1] else 0
        for i in reversed(range(0, n - 1)):
            rm[i] = rm[i + 1] + 1 if ratings[i] > ratings[i + 1] else 0
        for i in range(n):
            res += 1 + max(lm[i], rm[i])
        return res

def main():
    solver = Solution()
    ratings = [1, 2, 2]
    print(solver.candy(ratings))

if __name__ == '__main__':
    main()
